docs/source/triggers/index.rst

```diff
@@ -125,159 +111,316 @@ Like ``Scenario`` or ``Update``, you can use ``do`` notation and
+  message was sent by it and we don't need to reply to ourselves, so we do
+  nothing.
+- Finally, if we had not found any message at all, and thus ``lastMessage`` was
+  ``None``, we have nothing to do.
+
+Deduplication
+=============
+
+Daml Triggers react to many things, and it's usually important to make sure
+that the same command is not sent mutiple times.
+
+For example, in our ``autoReply`` chatbot above, the rule will be triggered not
+only when we receive a message, but also when we send one, as well as when we
+follow a user or get followed by a user, and when we stop following a user or a
+user stops following us.
+
+It's easy to imagine a sequence of events that would make a naive trigger
+implementation send too many messages. For example:
+
+- ``alice`` sends ``"hi"``, so the trigger runs and sends an ``exercise`` command.
+- _Before_ the ``exercise`` command is fully processed, ``carol`` follows
+  ``bob``, which triggers the rule again. The state of all the ``Message``
+  contracts ``bob`` can see has not changed, so the rule might send the
+  response to ``alice`` again.
+
+We obviously don't want that to happen, as it would likely prevent us from
+passing that Turing test we were after.
+
+Triggers offer a few features to help users manage that. Possibly the simplest
+one is the ``dedup*`` family of ledger operations. When using those, the
+trigger runner will keep track of the commands currently sent and prevent
+sending the exact same command again. In the above example, the trigger would
+see that, when ``carol`` follows ``bob`` and the rule runs ``dedupExercise``,
+there is already an Exercise command in flight with the exact same value, in
+this case same message, same sender and same receiver.
+
+Note that, if instead the in-between event is ``alice`` following ``carol``,
+this simple deduplication mechanism might not work as expected: because the
+``User`` contract ID for ``alice`` would have changed, the new command is not
+the same as the in-flight one and thus a second ``SendMessage`` exercise would
+be sent to the ledger.
+
+Similarly, if ``alice`` sends a second message quickly after the first one,
+this deduplication would prevent it, because the "response" does not have any
+reference to which message it's responding to. This may or may not be what we
+want.
+
+If this simple deduplication is not suited to your use-case, you have two other
+tools at your disposal. The first one is the ``getCommandsInflight`` action
+(`doc <https://docs.daml.com/triggers/api/Daml-Trigger.html#function-daml-trigger-getcommandsinflight-32524>`__),
+which returns all of the commands this instance of the trigger runner has sent
+and that have not yet been resolved (i.e. either committed or failed). You can
+then build your own logic based on this list, the ACS, and possibly your own
+trigger state.
+
+The last tool you have at your disposal is the second argument to the ``emitCommands`` action
```

Comment:
I’d swap these. `emitCommands` is pretty common. Writing your own logic based on commands in flight is not.

Reply:
Done.